pkg/sql/codegen/attribute/checker.go

```diff
@@ -17,87 +17,90 @@ import (
-func Float32RangeChecker(lower, upper *float32, includeLower, includeUpper bool) func(interface{}) error {
-
-	checker := func(e interface{}) error {
-		f, ok := e.(float32)
-		if !ok {
-			return fmt.Errorf("expected type float32, received %T", e)
-		}
-
-		// NOTE(tony): nil means no boundary
-		if lower != nil {
-			if includeLower && !(*lower <= f) {
-				return fmt.Errorf("range check %v <= %v failed", *lower, f)
-			}
-			if !includeLower && !(*lower < f) {
-				return fmt.Errorf("range check %v < %v failed", *lower, f)
+func Float32RangeChecker(lower, upper float32, includeLower, includeUpper bool) func(interface{}) error {
+	return func(attr interface{}) error {
+		if f, ok := attr.(float32); ok {
+			e := Float32LowerBoundChecker(lower, includeLower)(f)
+			if e == nil {
+				e = Float32UpperBoundChecker(upper, includeUpper)(f)
 			}
+			return e
 		}
+		return fmt.Errorf("expected type float32, received %T", attr)
+	}
+}
 
-		// NOTE(tony): nil means no boundary
-		if upper != nil {
-			if includeUpper && !(f <= *upper) {
-				return fmt.Errorf("range check %v <= %v failed", f, *upper)
-			}
-			if !includeUpper && !(f < *upper) {
-				return fmt.Errorf("range check %v < %v failed", f, *upper)
+// Float32LowerBoundChecker returns a range checker that only checks the lower bound.
+func Float32LowerBoundChecker(lower float32, includeLower bool) func(interface{}) error {
+	return func(attr interface{}) error {
+		if f, ok := attr.(float32); ok {
+			if f > lower || includeLower && f == lower {
+				return nil
 			}
+			return fmt.Errorf("range check %v <%v %v failed", lower, equalSign[includeLower], f)
 		}
-
-		return nil
+		return fmt.Errorf("expected type float32, received %T", attr)
 	}
-
-	return checker
 }
 
-func newInt(i int) *int {
-	return &i
+// Float32UpperBoundChecker returns a range checker that only checks the upper bound.
+func Float32UpperBoundChecker(upper float32, includeUpper bool) func(interface{}) error {
+	return func(attr interface{}) error {
+		if f, ok := attr.(float32); ok {
+			if f < upper || includeUpper && f == upper {
+				return nil
+			}
+			return fmt.Errorf("range check %v >%v %v failed", upper, equalSign[includeUpper], f)
+		}
+		return fmt.Errorf("expected type float32, received %T", attr)
+	}
 }
 
-// IntRangeChecker is a helper function to generate range checkers on attribute.
+// IntRangeChecker is a helper function to generate range checkers on attributes.
 // lower/upper indicates the lower bound and upper bound of the attribute value.
-// If lower/upper is nil, it means no boundary.
 // includeLower/includeUpper indicates the inclusion of the bound.
-func IntRangeChecker(lower, upper *int, includeLower, includeUpper bool) func(interface{}) error {
-	checker := func(e interface{}) error {
-		i, ok := e.(int)
-		if !ok {
-			return fmt.Errorf("expected type float32, received %T", e)
+func IntRangeChecker(lower, upper int, includeLower, includeUpper bool) func(interface{}) error {
```

Comment:
I think lower and upper should be a pointer, we can use `nil` to represent no-bounds: `(1, nil, true, false)` => `[1, ~)`

Reply:
This change uses LowerBoundChecker and UpperboundChecker to represent infinity bounds. The two functions behave similarly to the builtin operators `>=`, `<` etc. As a result, we don't need `nil` anymore.